SparseGrids/tsgGridFourier.cpp

```diff
@@ -0,0 +1,574 @@
+    delete[] basisFuncs;
+}
+
+double* GridFourier::getQuadratureWeights() const {
+    int num_points = getNumPoints();
+    double *w = new double[num_points]; 
+    getQuadratureWeights(w);
+    return w;
+}
+
+void GridFourier::getQuadratureWeights(double weights[]) const{
+    
+    /* 
+     * When integrating the Fourier series on a tensored grid, all the 
+     * nonzero modes vanish, and we're left with the normalized Fourier 
+     * coeff for e^0 (sum of the data divided by number of points)
+     */
+    
+    int num_points = getNumPoints();
+    std::fill(weights, weights+num_points, 0.0);
+    for(int n=0; n<active_tensors->getNumIndexes(); n++) {
+        const int *levels = active_tensors->getIndex(n);
+        int num_tensor_points = 1;
+        for(int j=0; j<num_dimensions; j++) {
+            num_tensor_points *= wrapper->getNumPoints(levels[j]);
+        }
+        for(int i=0; i<num_tensor_points; i++) {
+            weights[tensor_refs[n][i]] += ((double) active_weights[n])/((double) num_tensor_points);
+        }
+    }
+}
+
+void GridFourier::evaluate(const double x[], double y[]) const{
+    std::complex<double> *w = getBasisFunctions(x);
+    TasBLAS::setzero(num_outputs, y);
+    for(int k=0; k<num_outputs; k++){
+        for(int i=0; i<points->getNumIndexes(); i++){
+            y[k] += (w[i] * fourier_coefs[i*num_outputs+k]).real();
+        }
+    }
+    delete[] w;
+}
+
+void GridFourier::evaluateBatch(const double x[], int num_x, double y[]) const{
+    #pragma omp parallel for
+    for(int i=0; i<num_x; i++){
+        evaluate(&(x[((size_t) i) * ((size_t) num_dimensions)]), &(y[((size_t) i) * ((size_t) num_outputs)]));
+    }
+}
+
+void GridFourier::evaluateFastCPUblas(const double x[], double y[]) const{
+    #ifdef Tasmanian_ENABLE_BLAS
+    std::complex<double> *w = getBasisFunctions(x);
+    std::complex<double> *y_tmp = new std::complex<double>[num_outputs];
+    TasBLAS::zgemv(num_outputs, points->getNumIndexes(), fourier_coefs, w, y_tmp);
+    
+    #pragma omp parallel for
+    for(int i=0; i<num_outputs; i++) {
+        y[i] = y_tmp[i].real();
+    }
+    
+    delete[] w;
+    delete[] y_tmp;
+    #else
+    evaluate(x,y);
+    #endif // Tasmanian_ENABLE_BLAS
+}
+
+void GridFourier::evaluateFastGPUcublas(const double x[], double y[], std::ostream*) const{
+    evaluateFastCPUblas(x,y);
+}
+
+void GridFourier::evaluateFastGPUcuda(const double x[], double y[], std::ostream*) const{
+    evaluateFastCPUblas(x,y);
+}
+
+void GridFourier::evaluateFastGPUmagma(const double x[], double y[], std::ostream*) const{
+    evaluateFastCPUblas(x,y);
+}
+
+void GridFourier::evaluateBatchCPUblas(const double x[], int num_x, double y[]) const{
+    #ifdef Tasmanian_ENABLE_BLAS
+    int num_points = points->getNumIndexes();
+    std::complex<double> *y_tmp = new std::complex<double>[num_outputs * num_x];
+    std::complex<double> *weights = new std::complex<double>[num_points * num_x];
+    #pragma omp parallel for
```

Comment:
* isn't this the same as: `evaluateHierarchicalFunctions()`

Reply:
`evaluateHierarchicalFunctions()` gives the weights as an array of `double`s (real and imag parts), rather than complex numbers. Should I park this code in an overloaded function that gives the complex weights?

Reply:
Here you have the exact same loop as in `evaluateHierarchicalFunctions()`, just call the function, why write the loop twice?

Reply:
Done